domains/client/relayer/src/lib.rs

```diff
@@ -476,31 +479,75 @@ where
         .map_err(Error::UnableToSubmitCrossDomainMessage)
     }
 
-    fn last_relayed_block_key(domain_id: DomainId) -> Vec<u8> {
-        (b"message_relayer_last_processed_block_of_domain", domain_id).encode()
+    fn relayed_blocks_at_number_key(domain_id: DomainId, number: NumberFor<Block>) -> Vec<u8> {
+        (
+            b"message_relayer_processed_block_of_domain",
+            domain_id,
+            number,
+        )
+            .encode()
     }
 
-    fn fetch_last_relayed_block(
+    /// Takes number as tip and finds all the unprocessed blocks including the tip.
+    fn fetch_unprocessed_blocks_until(
         client: &Arc<Client>,
         domain_id: DomainId,
-    ) -> Option<NumberFor<Block>> {
-        let encoded = client
-            .get_aux(&Self::last_relayed_block_key(domain_id))
-            .ok()??;
+        number: NumberFor<Block>,
+        hash: Block::Hash,
+    ) -> Vec<(NumberFor<Block>, Block::Hash)> {
+        let mut blocks_to_process = vec![];
+        let (mut number_to_check, mut hash_to_check) = (number, hash);
+        while !Self::fetch_blocks_relayed_at(client, domain_id, number_to_check)
+            .contains(&hash_to_check)
+        {
+            blocks_to_process.push((number_to_check, hash_to_check));
+            hash_to_check = match client.header(hash_to_check).ok().flatten() {
+                Some(header) => *header.parent_hash(),
+                None => {
+                    error!(target: LOG_TARGET,
+                        "Failed to fetch header for hash {hash_to_check:?} for domain {domain_id:?}");
+                    return blocks_to_process;
+                }
+            };
+            number_to_check = match number_to_check.checked_sub(&One::one()) {
+                None => return blocks_to_process,
```

Comment:
This `return` doesn't make much sense to me either, but at the same time we can use `.expect()` here since it is not possible for parent block to be present and `checked_sub()` to return `None` at the same time.